pkg/apis/seedmanagement/v1alpha1/types_managedseed.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenletconfigv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list object metadata.
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot is the Shoot that will be registered as Seed.
+	// +optional
+	Shoot *Shoot `json:"shoot,omitempty" protobuf:"bytes,1,opt,name=shoot"`
+	// Seed describes the Seed that will be registered.
+	// Either Seed or Gardenlet must be specified.
+	// +optional
+	Seed *SeedTemplateSpec `json:"seed,omitempty" protobuf:"bytes,2,opt,name=seed"`
```

Comment:
Can we rather name this one `SeedTemplate`? I think, this can help to avoid confusion here.
```suggestion
	SeedTemplate *SeedTemplateSpec `json:"seedTemplate,omitempty" protobuf:"bytes,2,opt,name=seed"`
```

Reply:
There was similar feedback from @rfranzke during the review of GEP-13 and there I explained why I believe `seed` is better than `seedTemplate` here, see https://github.com/gardener/gardener/pull/3017#discussion_r526167508:

> I decided to rename it to `seed` because it's shorter, and because unlike template in Kubernetes resources (e.g. `ReplicaSet`) it's not only used to create new objects. Instead, any changes to the `seed` section will be reflected to the actual Seed if it exists.

I can still rename it, but I thought we have reached an agreement on this.

Reply:
Ok, didn't see that conversation on the GEP, but I got your motivation now.
My motivation to use `seedTemplate` originally was to make the difference between `shoot` and `seed` clear, as the referenced `shoot` has to exist already, while the `seed` will be created/reconciled (basically from the specified template).
Though, I don't have a strong preference here.

> I thought we have reached an agreement on this.

Actually, I perceived it this way, that the GEP was the main guideline how the described features will look like from a higher-level perspective and there's still room for fine-grained discussions like this one during the implementation phase :)

Reply:
OK, I will rename it to `SeedTemplate`, as I see this is what you and @rfranzke tend to prefer. I also don't have a strong preference.